feature_engine/creation/decision_tree.py

```diff
@@ -0,0 +1,328 @@
+                    "for output_features and {len(variables)} for the length "
+                    "of variables."
+                )
+
+        if isinstance(output_features, list):
+            if (
+                    max(output_features) > len(variables)
+                    or not all(
+                            isinstance(feature, int) for feature in output_features
+                        )
+            ):
+                raise ValueError(
+                    "output_features must be a list solely comprised of integers "
+                    "and the maximum integer cannot be greater than the length of "
+                    f"variables. Got {output_features} for output_features and "
+                )
+
+        if isinstance(output_features, tuple):
+            num_combos = 0
+            for n in range(1, len(variables) + 1):
+                num_combos += len(list(combinations(variables, n)))
+            if (
+                    not all(
+                        isinstance(feature, (str, tuple)) for feature in output_features
+                    )
+                    or len(output_features) > num_combos
+            ):
+                raise ValueError(
+                    "output_features must a tuple solely comprised of tuples and "
+                    f"the maximum number of tuples cannot exceed {num_combos}. Got "
+                    f"{output_features} instead."
+                )
+
+        if not isinstance(regression, bool):
+            raise ValueError(
+                f"regression must be a boolean value. Got {regression} instead."
+            )
+
+        if not isinstance(max_depth, int):
+            raise ValueError(
+                f"max_depth must be an integer. Got {max_depth} instead."
+            )
+
+        super().__init__(missing_value, drop_original)
+        self.variables = variables
+        self.output_features = output_features
+        self.regression = regression
+        self.max_depth = max_depth
+
+    def fit(self, X: pd.DataFrame, y: pd.Series):
+        """
+        The transformer learns the target variable values associated with
+        the user-provided features using a decision tree.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The training input samples. Can be the entire dataframe, not just
+            the variables to transform.
+
+        y: pandas Series or np.array = [n_samples,]
+            The target variable that is used to train the decision tree.
+        """
+        # common checks and attributes
+        # TODO: We don't need to check_variables b/c BaseCreation fit() performs action
+        X = super().fit(X, y)
+        self._check_dependent_variable_not_fitted_by_estimator(X, y)
+
+        self.variable_combinations_ = self._create_variable_combinations()
+        self.variable_combination_indices_ = {}
```

Comment:
what is the difference between these 2? if called indeces I expect integers, but I think these are variable names?

Reply:
`self.variable_combinations` is a **list of lists** of all the variable combinations to be used in the decision tree based on the value of `output_features`.

`self.variable_combination_indices_` is a **dictionary** with a key that has a unique estimator id and a value that is one of the lists from `self.variable_combinations`.

Example:
```
{'estimator_0': ['Age'],
 'estimator_1': ['Marks'],
 'estimator_2': ['Avg_5k_run_minutes'],
 'estimator_3': ['Height_cm'],
 'estimator_4': ['Age', 'Marks'],
 'estimator_5': ['Age', 'Avg_5k_run_minutes'],
 'estimator_6': ['Age', 'Height_cm'],
 'estimator_7': ['Marks', 'Avg_5k_run_minutes'],
 'estimator_8': ['Marks', 'Height_cm'],
 'estimator_9': ['Avg_5k_run_minutes', 'Height_cm'],
 'estimator_10': ['Age', 'Marks', 'Avg_5k_run_minutes'],
 'estimator_11': ['Age', 'Marks', 'Height_cm'],
 'estimator_12': ['Age', 'Avg_5k_run_minutes', 'Height_cm'],
 'estimator_13': ['Marks', 'Avg_5k_run_minutes', 'Height_cm']}
```

`self._fitted_estimators_` is a dictionary with the same keys and values as fitted decision trees. The keys/indices enable the mapping between the variable combinations and their corresponding fitted decision trees.

For some reason, I wanted to use a dictionary and a list/iterable can't be a key. However, I'm now thinking that we merge `self.variables_combination_indices_` and `self._fitted_estimators' into a list of tuples. Each tuple is comprised of 2 elements: a variable combination and its corresponding fitted decision tree. For now, let's call the merged variable `variable_combos_and_fitted_estimators`.

The goal of `variable_combos_and_fitted_estimators` is to ensure that DecisionTreeFeatures uses the correct variables when using the fitted estimator to predict new values in `transform()`.

Reply:
- I think having variable_combinations_ and variable_combinaction_indexes_ is reduntant.  
- Thinking out loud: can we have an attribute called `output_features_` that contains a tuple of tuples with the variable combinations (the combos). And then the hidden attribute `_estimators` (no need to add `fitted`) with a list of estimators fitted for those variable groups?

That might simplify things and also shorten the name of the attributes, that always helps :p